File: binarysearchtree_1.cpp
//initialite root node and loop until queue is empty . 
//remove first node and find mid
//create new node with middle node and set left child to dequeue node left child if present

#include <iostream>
#include <queue>
#include <vector>

using namespace std;

struct Node{
    int val; 
    Node* left; 
    Node* right; 
};

Node* sortedArrayToBST(vector<int>&nums){
    if (nums.empty()){
        return NULL;
    }

    int n = nums.size();
    int mid = n/2; 
    Node* root = new Node{nums[mid],NULL,NULL};
    //create queue    
    queue<pair<Node*, pair<int, int>>> q; 
    //push root to queue
    q.push({root, {0,mid-1}});
    q.push({root,{mid+1,n-1}});

    while (!q.empty()){
        //pull first element from the queue
        pair<Node*, pair<int, int> > curr=q.front();
        q.pop(); 

        Node* parent = curr.first;
        int left = curr.second.first; 
        int right = curr.second.second;

        //check if parent node is not null; 

        if (left<=right &&parent!=NULL){
            int mid = (left + right)/2; 
            Node* child = new Node{ nums[mid], NULL,NULL};

            if (nums[mid]<parent->val){
                parent->left = child; 
            }else{
                parent->right = child ; 
            }

            q.push({child,{left,mid-1}});
            q.push({child,{mid+1,right}});

        }
    }
    return root; 
}

void printBST(Node* root){
    if (root ==NULL)
        return ; 
    cout<<root->val<<" ";
    printBST(root->left);
    printBST(root->right);
}

int main(){
    vector<int> nums = {1,2,3,4,5,6,7};
    Node* root = sortedArrayToBST(nums);
    printBST(root);
    return 0 ; 
}

File: binarysearchtree_1.js
class Node {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

//function to create BST and return root
function sortedArrayToBST(arr) {
  if (arr.length == 0) {
    return null;
  }
  const mid = Math.floor(arr.lengh / 2);
  const root = new Node(arr[mid]);

  //initialize queue
  const q = [
    [root, [0, mid - 1]],
    [root, [mid + 1, arr.length - 1]],
  ];

  while (q.length > 0) {
    const [parent, [left, right]] = q.shift(); //shift removes the first element and shifts the other elements in its place

    //if there's elements to proces and parent node is not NULL;
    if (left <= right && parent != null) {
      const mid = Math.floor((left + right) / 2);
      const child = new Node(arr[mid]);
      console.log("left , right mid : ", left, right, mid);

      //set child node as left or right of parent

      if (arr[mid] < parent.val) {
        parent.left = child;
      } else {
        parent.right = child;
      }

      q.push([child, [left, mid - 1]]);
      q.push([child, [mid + 1, right]]);
    }
  }
  return root;
}

function printBST(root) {
  if (root === null) {
    return;
  }
  console.log(root.val + " ");
  printBST(root.left);
  printBST(root.right);
}

const nums = [1, 2, 3, 4, 5, 6, 7];
const root = sortedArrayToBST(nums);
printBST(root);
